rfcs/2023-11-08-font-loading-strategy.md

```diff
@@ -0,0 +1,211 @@
+    font-family: ${(theme) => theme.typography.fonts.family.text}
+  }
+`;
+```
+
+Live Example (POC): https://stackblitz.com/~/github.com/saurabhdaware/blade-self-hosted-font/tree/blade-import-css?file=index.tsx
+
+### React Native
+
+In React Native, people will continue to load the font the way they do right now (Download font files, link them to android and ios projects)
+
+Refer: https://blade.razorpay.com/?path=/docs/guides-installation--page#react-native-projects
+
+## Detailed Design
+
+### Criteria for font loading strategy
+
+Our recommended font loading strategy should find a good balance between following criteria
+
+- **Performance**
+  Should load fonts fast enough without impacting performance on rest of the app.
+
+- **DX**
+  Should not take significant amount of time and effort on consumer's end. Should be easy to load fonts.
+
+### Explored Solutions
+
+#### 1. Self Hosted: Manual Method
+
+- Download font files from respective sites
+- Define font-face in your CSS
+
+**Pros:**
+
+- Since consumers know the final font URL, they can be preloaded
+- Self-hosted so does not require additional DNS connection and thus fonts load faster
+
+**Cons:**
+
+- Consumers need to make sure fallbacks are set properly for browsers
+- Consumers need to make sure the place where fonts are hosted are correctly cached on CDN
+- Consumers need to make sure the correct lightweight fonts are used with correct configurations
+- Time consuming method to implement
+
+<details>
+<summary>Example</summary>
+
+```css
+@font-face {
+  font-family: 'TasaOrbiter';
+  src: url('/fonts/TASAOrbiterVF.woff2') format('woff2-variations');
+  font-weight: 125 950;
+  font-stretch: 75% 125%;
+  font-style: normal;
+}
+
+@font-face {
+  font-family: 'Inter';
+  src: url('/fonts/Inter-Regular.ttf') format('truetype');
+  font-weight: normal;
+  font-style: normal;
+}
+
+@font-face {
+  font-family: 'Inter';
+  src: url('/fonts/Inter-SemiBold.ttf') format('truetype');
+  font-weight: 600;
+  font-style: normal;
+}
+```
+
+CSR apps can also add preloads to fonts to load them faster
+
+```html
+<link rel="preload" href="/fonts/TASAOrbiterVF.woff2" as="font" type="font/woff2" crossorigin />
+<link rel="preload" href="/fonts/Inter-Regular.ttf" as="font" type="font/ttf" crossorigin />
+<link rel="preload" href="/fonts/Inter-SemiBold.ttf" as="font" type="font/ttf" crossorigin />
+```
+
+Example: https://stackblitz.com/~/github.com/saurabhdaware/blade-self-hosted-font
+
+</details>
+
+<!-- #### Difference between loading fonts without preload and with preload
+
+https://github.com/razorpay/blade/assets/30949385/b2b12cbe-2c93-4af9-bc7f-7b94ed36b90b
+
+https://github.com/razorpay/blade/assets/30949385/d468cfc1-c1a9-44ad-9d85-a3ea37320ce7 -->
+
+#### 2. Self Hosted: Using CSS Imports from Blade (✅ Preferred)
+
+Similar to how we load Lato font from `@fontsource/lato` right now, Consumers add import from `@razorpay/blade/fonts.css`
+
+```js
+import '@razorpay/blade/fonts.css';
```

Comment:
wouldn't this be render blocking when the consumers are using it as this will be at the beginning of the file?

Reply:
I guess bundler will load this css file and add it to HTML like a normal `<link rel="stylesheet"` so its same as normally linking CSS only
<img width="519" alt="image" src="https://github.com/razorpay/blade/assets/30949385/3a5e3ec9-fe77-4d1f-82e7-56ccb92b496b">

If you mean render-blocking in sense how normal stylesheet links are render-blocking in HTML, don't think so we can do much about it.